### PR TITLE
WIP: Input field deprecation support

### PR DESCRIPTION
**Pull request by jturkel**
This PR implements the [RFC](https://github.com/graphql/graphql-spec/pull/525) for input field deprecation support. The RFC was moved to stage 2 at the last [working group](https://docs.google.com/document/d/1vw5zEOHVPBtspoWRVQ279NSo_JfWg7ZEdGKix4O-nco/edit#heading=h.w19vr9farz2x). I don't think we want to merge this PR until the corresponding graphql-js [PR](https://github.com/graphql/graphql-js/pull/1560) merges (or the RFC moves to the accepted state) but I wanted to open a WIP PR in case there's an early feedback on my implementation and to prevent anyone from doing duplicate work.

### CODE REVIEW

**Comment by Marthyn** [comment]
Thanks for this PR! Was just about to use it but then found out it's being worked on.

**Comment by rgraff** [comment]
Could really use this as well.

**Comment by jturkel** [comment]
@Marthyn / @rgraff - I wasn't in a rush to merge this since none of my clients currently leverage this information. Also the interaction with introspection is still undergoing some discussion. As written here and in graphql-js, marking an argument as deprecated will hide it from introspection queries (even if it's required) unless clients pass `includeDeprecated: true`. This is challenging for generic clients since they don't know if the server does or doesn't support argument deprecation so a two phase introspection process is required. All that said, I'd be happy to work on getting this merged sooner rather than later if you're going to use it now.

**Comment by kroehre** [comment]
@jturkel I have a case for this as well, would love to see this merged!

**Comment by jturkel** [comment]
@rmosolgo - I've rebased this branch to resolve the merge conflict. We're going to hold off using this feature until graphql-js merges/releases https://github.com/graphql/graphql-js/pull/1560 but I'm totally fine merging the graphql-ruby PR as-is if other folks want to use the feature sooner.

**Comment by kroehre** [comment]
@rmosolgo Any blockers here? I would love to use this feature. Thanks!

**Comment by rusterholz** [comment]
@jturkel Looks like that PR you linked to has been open for two years. Any chance that _this_ PR could get merged before that one?

**Comment by bilby91** [comment]
Looking forward on this feature!!

**Comment by jturkel** [comment]
Looks like the graphql-js change merged in https://github.com/graphql/graphql-js/pull/2733. The only change from what I implemented here is that you can't deprecate required arguments. I'll make that change tomorrow.

**Comment by rmosolgo** [comment]
:+1: Thanks for the update, @jturkel . Could you @-me after pushing those changes? Sorry I've been slow to merge, I'll go ahead and merge it when that's taken care of.

**Comment by jturkel** [comment]
@rmosolgo - I added validation that deprecated arguments aren't required so this should be ready to go. I made that change in a separate commit for easier review but I'd be happy to rebase/squash to keep the master commit history clean if you want.

**Comment by rmosolgo** [comment]
Awesome, thanks again for your work on this!

**Comment by rmosolgo** [comment]
Anyone who wants to give this a try before the next release can pull the gem from github: 

```ruby 
gem "graphql", github: "rmosolgo/graphql-ruby"
```

**Comment by Marthyn** [comment]
Awesome! Thank you!

File: spec/graphql/schema/build_from_definition_spec.rb [hunk 1]
```diff
@@ -743,10 +743,16 @@ def assert_schema_and_compare_output(definition)
   VALUE
 }
 
+input MyInput {
+  int: Int @deprecated(reason: "This is not the argument you're looking for")
```
Comment (rimian) [inline spec/graphql/schema/build_from_definition_spec.rb:747]: This is not the argument I'm looking for.

**Pull request merged**